paperqa/agents/task.py

```diff
@@ -487,3 +499,165 @@ def __len__(self) -> int:
+        evaluation_callback: (
+            Callable[[MultipleChoiceEvaluation | dict], Awaitable] | None
+        ) = None,
+        **kwargs,
+    ):
+        kwargs["query"] = question
+        kwargs["docs"] = Docs()
+        super().__init__(*args, **kwargs)
+
+        self.qid = qid
+        self.question = question
+        self.human_answer = human_answer
+        self.gt_doc_ids = gt_doc_ids
+        self.pairwise_eval_llm = pairwise_eval_llm
+        self._evaluation_callback = evaluation_callback
+
+    def extract_best_answer_index(self, text: str) -> int:
+        match = re.search(r"<rating>(\d+)</rating>", text)
+        return int(match.group(1)) if match else 0
+
+    async def pairwise_evaluation(
+        self, qid: str | UUID, question: str, pqa_answer: str, human_answer: str
+    ) -> dict:
+
+        paper_search_ids = [int(doc.docname) for doc in self.state.docs.docs.values()]
+
+        pairwise_eval_llm = LiteLLMModel(name=self.pairwise_eval_llm)
+        pqa_answer = strip_citations(pqa_answer)
+
+        pqa_answer_index = 1 if random.random() < 0.5 else 2  # noqa: PLR2004
+        data = {
+            "question": question,
+            "answer1": pqa_answer if pqa_answer_index == 1 else human_answer,
+            "answer2": human_answer if pqa_answer_index == 1 else pqa_answer,
+        }
+
+        result = await pairwise_eval_llm.run_prompt(
+            prompt=lfrqa_prompt_template,
+            data=data,
+            system_prompt=lfrqa_system_prompt,
+        )
+
+        best_answer_index = self.extract_best_answer_index(result.text)
+        if best_answer_index == pqa_answer_index:
+            winner, reward = "paperqa", self._rewards["win"]
+        elif best_answer_index != 0:
+            winner, reward = "human", self._rewards["lose"]
+        else:
+            winner, reward = "tie", self._rewards["tie"]
+
+        return {
+            "llm": self._settings.llm,
+            "evaluator_llm": self.pairwise_eval_llm,
+            "qid": qid,
+            "question": question,
+            "pqa_answer": pqa_answer,
+            "human_answer": human_answer,
+            "winner": winner,
+            "paper_search_ids": paper_search_ids,
+            "gt_doc_ids": self.gt_doc_ids,
+            "pqa_answer_was_answer_1": pqa_answer_index == 1,
+            "complete_evaluator_response": result.text,
+            "reward": reward,
+        }
+
+    async def step(
+        self, action: ToolRequestMessage
+    ) -> tuple[Messages, float, bool, bool]:
+        messages, reward, done, truncated = await super().step(action)
+        if not done:
+            return messages, reward, done, truncated
+
+        evaluation = await self.pairwise_evaluation(
+            self.qid, self.question, self.state.session.answer, self.human_answer
+        )
+
+        if evaluation_callback := self._evaluation_callback:
+            await evaluation_callback(evaluation)
```

Comment:
Can we blow up here if the `evaluation` is a `MultipleChoiceEvaluation`?

Reply:
Yeah, wdyt of:
```python
raise ValueError(
    f"Evaluation in {type(self).__name__} must be input type"
    f" {dict.__name__}, we received"
    f" type {type(evaluation).__name__}."
    f" Note that {type(self).__name__} is a subclass of"
    f" {GradablePaperQAEnvironment.__name__}."
)
```

Reply:
Discussed in person, pretty good but remember to mention the "why" -- that's what helps users